openedx_proversity_reports/reports/time_spent_report.py

```diff
@@ -0,0 +1,239 @@
+            ga_credentials, scopes=ANALYTICS_API_SCOPES
+        )
+
+        # Build the service object.
+        analytics = build('analyticsreporting', 'v4', credentials=credentials, cache_discovery=False)
+
+        return analytics
+
+
+    def get_analytics_report(analytics):
+        """
+        Queries the Analytics Reporting API V4.
+
+        Args:
+            analytics: An authorized Analytics Reporting API V4 service object.
+        Returns:
+            The Analytics Reporting API V4 response or None if analytics view id is not provided.
+        """
+        course_expression = '{}{}'.format(course_id, COURSEWARE_URL_PREFIX)
+        analytics_view_id = getattr(settings, 'OPR_GOOGLE_ANALYTICS_VIEW_ID', None)
+
+        if not analytics_view_id:
+            logger.error('Google Analytics view id has not been provided.')
+            raise GoogleAnalyticsCredentialsError('Google Analytics view id has not been provided.')
+
+        return analytics.reports().batchGet(
+            body={
+                'reportRequests': [{
+                    'viewId': analytics_view_id,
+                    'dateRanges': [
+                        {'startDate': '7DaysAgo', 'endDate': 'yesterday'},
+                    ],
+                    'metrics': [
+                        {'expression': 'ga:pageviews'},
+                        {'expression': 'ga:avgTimeOnPage'},
+                    ],
+                    'dimensions': [{'name': 'ga:pagePath'}],
+                    "dimensionFilterClauses": [{
+                        "filters": [{
+                            "dimensionName": 'ga:pagePath',
+                            "not": False,
+                            "operator": 'PARTIAL',
+                            "expressions": [
+                                course_expression,
+                            ],
+                            "caseSensitive": False
+                        }]
+                    }]
+                }]
+            }
+        ).execute()
+
+
+    def parse_response(response):
+        """
+        Parses the Analytics Reporting API V4 response.
+
+        Args:
+            response: An Analytics Reporting API V4 response.
+        """
+        data = []
+        for report in response.get('reports', []):
+            for row in report.get('data', {}).get('rows', []):
+                dimensions = row.get('dimensions', [])
+                metric_values = row.get('metrics', [])
+                # Only take the first list value, since only one date range is requested.
+                dimension_values = metric_values[0].get('values', [])
+
+                if dimension_values:
+                    # dimensions[0], since only one dimension is requested.
+                    # dimension_values[0], it's the ga:pageviews number.
+                    # dimension_values[0], it's the ga:avgTimeOnPage time number in seconds.
+                    data.append({
```

Comment:
possible IndexError could be raised

Reply:
Yeap, a try except block will be applied, but if there is an error, we should not be resilient, since this information is required to report generation. @andrey-canon